File: src/main/java/eladkay/quaritum/common/rituals/works/CircleOfTheFinalMomentWork.java
package eladkay.quaritum.common.rituals.works;

import eladkay.quaritum.api.rituals.IDiagram;
import eladkay.quaritum.api.rituals.IWork;
import eladkay.quaritum.api.rituals.PositionedBlock;
import eladkay.quaritum.api.rituals.PositionedBlockChalk;
import eladkay.quaritum.common.networking.FancyParticlePacket;
import eladkay.quaritum.common.networking.NetworkHelper;
import eladkay.quaritum.common.rituals.diagrams.CircleOfTheFinalMomentDiagram;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.ITickable;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.List;

public class CircleOfTheFinalMomentWork implements IWork {
    @Nonnull
    @Override
    public String getUnlocalizedName() {
        return "circleofthefinalmoment";
    }

    @Override
    public boolean updateTick(@Nonnull World world, @Nonnull BlockPos pos, @Nonnull TileEntity tile, long ticksExisted) {
        for (BlockPos pos0 : BlockPos.getAllInBox(new BlockPos(-4, 1, -4).add(pos), new BlockPos(4, 8, 4).add(pos)))
            if(world.getTileEntity(pos0) != null && world.getTileEntity(pos0) instanceof ITickable)
                for(int i = 0; i < 4; i++) ((ITickable) world.getTileEntity(pos0)).update();
        NetworkHelper.tellEveryoneAround(new FancyParticlePacket(pos.getX() + 0.25, pos.getY(), pos.getZ() + 0.25, 50), world.provider.getDimension(), pos, 32);
        return ticksExisted < 1200;
    }

    @Override
    public boolean initialTick(@Nonnull World world, @Nonnull BlockPos pos, @Nonnull TileEntity tile, @Nullable EntityPlayer player) {
        return true;
    }

    @Override
    public boolean canRitualRun(@Nonnull World world, @Nonnull BlockPos pos, @Nonnull TileEntity tile) {
        return true;
    }

    @Override
    public void buildPositions(@Nonnull List<PositionedBlock> chalks) {
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-4, 1, -7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-3, 1, -7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-2, 1, -7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-1, 1, -7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(0, 1, -7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(1, 1, -7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(2, 1, -7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(3, 1, -7)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-6, 1, -6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-5, 1, -6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-4, 1, -6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(3, 1, -6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(4, 1, -6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(5, 1, -6)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(6, 1, -6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-6, 2, -6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(6, 2, -6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-6, 3, -6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(6, 3, -6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(-6, 4, -6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(6, 4, -6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-6, 1, -5)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-5, 1, -5)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(0, 1, -5), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(5, 1, -5)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(6, 1, -5)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(0, 2, -5), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(0, 3, -5), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(0, 4, -5), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(0, 5, -5)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-6, 1, -4)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(-4, 1, -4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-2, 1, -4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-1, 1, -4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(0, 1, -4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(1, 1, -4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(2, 1, -4)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(4, 1, -4)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(6, 1, -4)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-7, 1, -3)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-6, 1, -3)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-3, 1, -3)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-2, 1, -3)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(2, 1, -3)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(3, 1, -3)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(6, 1, -3)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(7, 1, -3)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-7, 1, -2)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-4, 1, -2)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-3, 1, -2)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(3, 1, -2)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(4, 1, -2)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(7, 1, -2)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-7, 1, -1)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-4, 1, -1)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(4, 1, -1)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(7, 1, -1)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-7, 1, 0)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-5, 1, 0), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-4, 1, 0)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(4, 1, 0)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(5, 1, 0), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(7, 1, 0)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-5, 2, 0), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(5, 2, 0), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-5, 3, 0), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(5, 3, 0), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-5, 4, 0), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(5, 4, 0), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(-5, 5, 0)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(5, 5, 0)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-7, 1, 1)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-4, 1, 1)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(4, 1, 1)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(7, 1, 1)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-7, 1, 2)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-4, 1, 2)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-3, 1, 2)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(3, 1, 2)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(4, 1, 2)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(7, 1, 2)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-7, 1, 3)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-6, 1, 3)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-3, 1, 3)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-2, 1, 3)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(2, 1, 3)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(3, 1, 3)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(6, 1, 3)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(7, 1, 3)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-6, 1, 4)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(-4, 1, 4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-2, 1, 4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(-1, 1, 4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(0, 1, 4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(1, 1, 4)));
        chalks.add(new PositionedBlockChalk(null, new BlockPos(2, 1, 4)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(4, 1, 4)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(6, 1, 4)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-6, 1, 5)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-5, 1, 5)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(0, 1, 5), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(5, 1, 5)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(6, 1, 5)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(0, 2, 5), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(0, 3, 5), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(0, 4, 5), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(0, 5, 5)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-6, 1, 6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-5, 1, 6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-4, 1, 6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-3, 1, 6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(3, 1, 6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(4, 1, 6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(5, 1, 6)));
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(6, 1, 6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-6, 2, 6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(6, 2, 6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(-6, 3, 6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlock(Blocks.OBSIDIAN.getDefaultState(), new BlockPos(6, 3, 6), null)); // minecraft:obsidian
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(-6, 4, 6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.RED, new BlockPos(6, 4, 6)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-3, 1, 7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-2, 1, 7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(-1, 1, 7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(0, 1, 7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(1, 1, 7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(2, 1, 7)));
        chalks.add(new PositionedBlockChalk(EnumDyeColor.YELLOW, new BlockPos(3, 1, 7)));
    }

    @Nullable
    @Override
    public IDiagram getDiagramCounterpart() {
        return new CircleOfTheFinalMomentDiagram();
    }
}

File: src/main/java/eladkay/quaeritum/common/rituals/works/CircleOfTheFinalMomentWork.kt
package eladkay.quaeritum.common.rituals.works

import eladkay.quaeritum.api.animus.EnumAnimusTier
import eladkay.quaeritum.api.rituals.IDiagram
import eladkay.quaeritum.api.rituals.IWork
import eladkay.quaeritum.api.rituals.PositionedBlock
import eladkay.quaeritum.api.rituals.PositionedBlockChalk
import eladkay.quaeritum.common.networking.FancyParticlePacket
import eladkay.quaeritum.common.networking.NetworkHelper
import eladkay.quaeritum.common.rituals.diagrams.CircleOfTheFinalMomentDiagram
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Blocks
import net.minecraft.item.EnumDyeColor
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.ITickable
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

class CircleOfTheFinalMomentWork : IWork {
    override fun getUnlocalizedName(): String {
        return "circleofthefinalmoment"
    }

    override fun updateTick(world: World, pos: BlockPos, tile: TileEntity, ticksExisted: Long): Boolean {
        for (pos0 in BlockPos.getAllInBox(BlockPos(-4, 1, -4).add(pos), BlockPos(4, 8, 4).add(pos)))
            if (world.getTileEntity(pos0) != null && world.getTileEntity(pos0) is ITickable) {
                for (i in 0..3) (world.getTileEntity(pos0) as ITickable).update()
                NetworkHelper.tellEveryoneAround(FancyParticlePacket(pos0.x + 0.25, pos0.y - 0.5, pos0.z + 0.25, 5), world.provider.dimension, pos, 32)
            }
        NetworkHelper.tellEveryoneAround(FancyParticlePacket(pos.x + 0.25, pos.y.toDouble(), pos.z + 0.25, 50), world.provider.dimension, pos, 32)
        return IDiagram.Helper.consumeAnimusForRitual(tile, true, 1, EnumAnimusTier.QUAERITUS)
        //ticksExisted < 1200;
    }

    override fun initialTick(world: World, pos: BlockPos, tile: TileEntity, player: EntityPlayer?): Boolean {
        return true
    }

    override fun canRitualRun(world: World, pos: BlockPos, tile: TileEntity): Boolean {
        return true
    }

    override fun buildPositions(chalks: MutableList<PositionedBlock>) {
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-3, 1, -7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-2, 1, -7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-1, 1, -7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(0, 1, -7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(1, 1, -7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(2, 1, -7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(3, 1, -7)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-6, 1, -6), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-5, 1, -6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-4, 1, -6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-3, 1, -6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(3, 1, -6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(4, 1, -6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(5, 1, -6)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(6, 1, -6), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-6, 2, -6), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(6, 2, -6), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-6, 3, -6), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(6, 3, -6), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(-6, 4, -6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(6, 4, -6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-6, 1, -5)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-5, 1, -5)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(0, 1, -5), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(5, 1, -5)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(6, 1, -5)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(0, 2, -5), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(0, 3, -5), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(0, 4, -5), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(0, 5, -5)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-6, 1, -4)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(-4, 1, -4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-2, 1, -4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-1, 1, -4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(0, 1, -4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(1, 1, -4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(2, 1, -4)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(4, 1, -4)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(6, 1, -4)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-7, 1, -3)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-6, 1, -3)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-3, 1, -3)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-2, 1, -3)))
        chalks.add(PositionedBlockChalk(null, BlockPos(2, 1, -3)))
        chalks.add(PositionedBlockChalk(null, BlockPos(3, 1, -3)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(6, 1, -3)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(7, 1, -3)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-7, 1, -2)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-4, 1, -2)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-3, 1, -2)))
        chalks.add(PositionedBlockChalk(null, BlockPos(3, 1, -2)))
        chalks.add(PositionedBlockChalk(null, BlockPos(4, 1, -2)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(7, 1, -2)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-7, 1, -1)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-4, 1, -1)))
        chalks.add(PositionedBlockChalk(null, BlockPos(4, 1, -1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(7, 1, -1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-7, 1, 0)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-5, 1, 0), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(null, BlockPos(-4, 1, 0)))
        chalks.add(PositionedBlockChalk(null, BlockPos(4, 1, 0)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(5, 1, 0), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(7, 1, 0)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-5, 2, 0), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(5, 2, 0), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-5, 3, 0), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(5, 3, 0), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-5, 4, 0), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(5, 4, 0), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(-5, 5, 0)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(5, 5, 0)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-7, 1, 1)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-4, 1, 1)))
        chalks.add(PositionedBlockChalk(null, BlockPos(4, 1, 1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(7, 1, 1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-7, 1, 2)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-4, 1, 2)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-3, 1, 2)))
        chalks.add(PositionedBlockChalk(null, BlockPos(3, 1, 2)))
        chalks.add(PositionedBlockChalk(null, BlockPos(4, 1, 2)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(7, 1, 2)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-7, 1, 3)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-6, 1, 3)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-3, 1, 3)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-2, 1, 3)))
        chalks.add(PositionedBlockChalk(null, BlockPos(2, 1, 3)))
        chalks.add(PositionedBlockChalk(null, BlockPos(3, 1, 3)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(6, 1, 3)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(7, 1, 3)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-6, 1, 4)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(-4, 1, 4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-2, 1, 4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(-1, 1, 4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(0, 1, 4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(1, 1, 4)))
        chalks.add(PositionedBlockChalk(null, BlockPos(2, 1, 4)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(4, 1, 4)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(6, 1, 4)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-6, 1, 5)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-5, 1, 5)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(0, 1, 5), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(5, 1, 5)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(6, 1, 5)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(0, 2, 5), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(0, 3, 5), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(0, 4, 5), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(0, 5, 5)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-6, 1, 6), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-5, 1, 6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-4, 1, 6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-3, 1, 6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(3, 1, 6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(4, 1, 6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(5, 1, 6)))
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(6, 1, 6), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-6, 2, 6), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(6, 2, 6), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(-6, 3, 6), null)) // minecraft:obsidian
        chalks.add(PositionedBlock(Blocks.OBSIDIAN.defaultState, BlockPos(6, 3, 6), null)) // minecraft:obsidian
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(-6, 4, 6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.RED, BlockPos(6, 4, 6)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-3, 1, 7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-2, 1, 7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(-1, 1, 7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(0, 1, 7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(1, 1, 7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(2, 1, 7)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.YELLOW, BlockPos(3, 1, 7)))
    }

    override fun getDiagramCounterpart(): IDiagram? {
        return CircleOfTheFinalMomentDiagram()
    }
}
